apollo-router-core/src/query_cache.rs

```diff
@@ -0,0 +1,107 @@
+use crate::prelude::graphql::*;
+use futures::lock::Mutex;
+use lru::LruCache;
+use std::collections::HashMap;
+use std::sync::Arc;
+use tokio::sync::broadcast;
+
+/// A cache for parsed GraphQL queries.
+#[derive(Debug)]
+pub struct QueryCache {
+    cached: Mutex<LruCache<String, Option<Arc<Query>>>>,
+    #[allow(clippy::type_complexity)]
+    wait_map: Mutex<HashMap<String, broadcast::Sender<(String, Option<Arc<Query>>)>>>,
+    cache_limit: usize,
+    schema: Arc<Schema>,
+}
+
+impl QueryCache {
```

Comment:
Can we have some basic unit tests for this?

Reply:
I propose we delay that for the PR that will refactor the caching thing as this should be done only once.